### PR TITLE
[TypeScript] Make types more strict in ra-core, part III

### PR DESCRIPTION
**Pull request by fzaninotto**
This PR finally turns `strictNullChecks` to `true` in `ra-core`, and solves the remaining 63 TS compilation errors on `ra-core`.

Refs #9622
Follows #9644, #9741 , #9743

## `useRecordContext` Returns `undefined` When No Record Is Available

The `useRecordContext` hook reads the current record from the `RecordContext`. This context may be empty (e.g. while the record is being fetched). The return type for `useRecordContext` has been modified to `Record | undefined` instead of `Record` to denote this possibility.

As a consequence, the TypeScript compilation of your project may fail if you don't check the existence of the record before reading it.

To fix this error, your code should handle the case where `useRecordContext` returns `undefined`:

```diff
const MyComponent = () => {
    const record = useRecordContext();
+   if (!record) return null;
    return (
        <div>
            <h1>{record.title}</h1>
            <p>{record.body}</p>
        </div>
    );
};
```

### CODE REVIEW

File: packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx [hunk 1]
```diff
@@ -229,60 +230,62 @@ export const Datagrid: FC<DatagridProps> = React.forwardRef((props, ref) => {
...
-                                rowStyle,
-                                selectedIds,
-                                isRowSelectable,
-                            },
-                            children
-                        )}
-                    </Table>
-                </div>
-            </DatagridRoot>
+            <OptionalResourceContextProvider value={resource}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx:233]: Stricter types revealed a bug: Using a Datagrid in controlled mode with an EditButton as a child used to fail silently (the EditButton link didn't work), because the resource set on the Datagrid was never propagated to the child. Now the failure is loud ans forces to add this ResourceContext.

File: packages/ra-core/src/form/useFormGroup.ts [hunk 2]
```diff
@@ -109,11 +110,13 @@ export const useFormGroup = (name: string): FormGroupState => {
...
 
     useEffect(() => {
+        if (!formGroups) return;
         // Whenever the group content changes (input are added or removed)
         // we must update its state
-        return formGroups.subscribe(name, () => {
+        const unsubscribe = formGroups.subscribe(name, () => {
             updateGroupState();
         });
+        return unsubscribe;
```
Comment (djhi) [inline packages/ra-core/src/form/useFormGroup.ts:119]: I don't think we have an `unsubscribe` method. We should add one

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.spec.tsx [hunk 3]
```diff
@@ -70,7 +70,7 @@ describe('<ReferenceArrayField />', () => {
                             record={{ id: 123, barIds: [1, 2] }}
                             reference="bar"
                         >
-                            <SingleFieldList>
+                            <SingleFieldList linkType={false}>
```
Comment (djhi) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.spec.tsx:73]: Why is this necessary? `linkType` is optional in `SingleFieldList` props

File: packages/ra-core/src/controller/list/useListParams.ts [hunk 4]
```diff
@@ -157,6 +157,11 @@ export const useListParams = ({
                 tempParams.current = queryReducer(query, action);
                 // schedule side effects for next tick
                 setTimeout(() => {
+                    if (!tempParams.current) {
+                        throw new Error(
+                            'Race condition in changeParams detected'
```
Comment (djhi) [inline packages/ra-core/src/controller/list/useListParams.ts:162]: Users might need more information to solve this

**Review by djhi — CHANGES_REQUESTED** [review]

**Review by adguernier — APPROVED** [review]

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.spec.tsx (hunk 3 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.spec.tsx:73]: Because the test doesn't define a `ResourceContext`, and `<SingleFieldList>` defaults to creating a link. It's not possible to create a link without a Resource. Before this PR, this caused an empty/invalid link. Now, it throws an error. To avoid this error, we must disable the ink or define a `ResourceContext`.

File: packages/ra-core/src/controller/list/useListParams.ts (hunk 4 above)
Comment (fzaninotto) [reply packages/ra-core/src/controller/list/useListParams.ts:162]: I agree, but this error should never occur (check the source code). To be honest, I added it just to please TypeScript.

File: packages/ra-core/src/form/useFormGroup.ts (hunk 2 above)
Comment (fzaninotto) [reply packages/ra-core/src/form/useFormGroup.ts:119]: Yes we do:

https://github.com/marmelab/react-admin/blob/131619a8bfca889968daab748f1993e25c763d54/packages/ra-core/src/form/FormGroupsProvider.tsx#L34-L38

File: packages/ra-core/src/controller/list/useListParams.ts (hunk 4 above)
Comment (fzaninotto) [reply packages/ra-core/src/controller/list/useListParams.ts]: Now that you mention it, throwing an error may break the app in legitimate cases (e.g. two successive changeParams calls). I've converted it to an early return.

**Review by djhi — APPROVED** [review]

File: packages/ra-core/src/core/CoreAdminContext.tsx [hunk 5]
```diff
@@ -187,9 +191,11 @@ React-admin requires a valid dataProvider function to work.`);
...
     const finalAuthProvider = useMemo(
         () =>
-            authProvider instanceof Function
-                ? convertLegacyAuthProvider(authProvider)
-                : authProvider,
+            authProvider
+                ? authProvider instanceof Function
+                    ? convertLegacyAuthProvider(authProvider)
+                    : authProvider
+                : defaultAuthProvider,
```
Comment (slax57) [inline packages/ra-core/src/core/CoreAdminContext.tsx:198]: Just found out this is a BC, because now all admins have a default non-null authProvider.
So this breaks code like the following:

```tsx
        <Root
            label={
                authProvider != null
                    ? 'ra.auth.user_menu'
                    : 'ra.configurable.customize'
            }
            icon={
                authProvider ? (
                    <Avatar src={identity.avatar} />
                ) : (
                    <SettingsIcon />
                )
            }
            {...props}
        />
```

Is there a way to avoid it?

Comment (slax57) [reply packages/ra-core/src/core/CoreAdminContext.tsx:198]: https://github.com/marmelab/react-admin/pull/9861 attempts to fix this BC
